core/trino-spi/src/main/java/io/trino/spi/block/DecimalAggregationAccumulatorBlock.java

```diff
@@ -0,0 +1,330 @@
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.spi.block.BlockUtil.checkArrayRange;
+import static io.trino.spi.block.BlockUtil.checkValidRegion;
+import static io.trino.spi.block.BlockUtil.compactArray;
+import static io.trino.spi.block.BlockUtil.countUsedPositions;
+
+public class DecimalAggregationAccumulatorBlock
+        implements Block
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(DecimalAggregationAccumulatorBlock.class).instanceSize();
+
+    private final int positionOffset;
+    private final int positionCount;
+    @Nullable
+    private final boolean[] valueIsNull;
+    private final long[] values;
+    @Nullable
+    private final long[] overflow;
+
+    private final long sizeInBytes;
+    private final long retainedSizeInBytes;
+    private final int numberOfNoOverflowValuesPerEntry;
+    private final int maxOffset;
+
+    public DecimalAggregationAccumulatorBlock(int positionCount, Optional<boolean[]> valueIsNull, long[] values, Optional<long[]> overflow, int numberOfNoOverflowValuesPerEntry)
+    {
+        this(0, positionCount, valueIsNull.orElse(null), values, overflow.orElse(null), numberOfNoOverflowValuesPerEntry);
+    }
+
+    DecimalAggregationAccumulatorBlock(int positionOffset, int positionCount, boolean[] valueIsNull, long[] values, long[] overflow, int numberOfNoOverflowValuesPerEntry)
+    {
+        if (positionOffset < 0) {
+            throw new IllegalArgumentException("positionOffset is negative");
+        }
+        this.positionOffset = positionOffset;
+        if (positionCount < 0) {
+            throw new IllegalArgumentException("positionCount is negative");
+        }
+        this.positionCount = positionCount;
+
+        this.numberOfNoOverflowValuesPerEntry = numberOfNoOverflowValuesPerEntry;
+
+        // offset used for overflow
+        this.maxOffset = 8 * numberOfNoOverflowValuesPerEntry;
+
+        if (values.length - (positionOffset * numberOfNoOverflowValuesPerEntry) < positionCount * numberOfNoOverflowValuesPerEntry) {
+            throw new IllegalArgumentException("values length is less than positionCount");
+        }
+        this.values = values;
+
+        if (valueIsNull != null && valueIsNull.length - positionOffset < positionCount) {
+            throw new IllegalArgumentException("isNull length is less than positionCount");
+        }
+        this.valueIsNull = valueIsNull;
+
+        if (overflow != null && overflow.length - positionOffset < positionCount) {
+            throw new IllegalArgumentException("overflow length is less than positionCount");
+        }
+        this.overflow = overflow;
+
+        sizeInBytes = (numberOfNoOverflowValuesPerEntry + Byte.BYTES + Long.BYTES) * (long) positionCount;
```

Comment:
you should multiply `numberOfNoOverflowValuesPerEntry` with `Long.BYTES`

nit: this probably should take into account that `overflow` might be null